src/devices/machine/plotter.cpp

```diff
@@ -0,0 +1,268 @@
+// license:BSD-3-Clause
+// copyright-holders: Fabio Dalla Libera
+/*
+  Based on bitmap_printer.cpp
+
+  TODO: - save snapshots a vector graphics
+  - optimize copybitmapregion
+
+
+*/
+
+#include "plotter.h"
+#include "emuopts.h"
+#include "fileio.h"
+#include "png.h"
+
+static inline int32_t clip(int32_t v,int32_t minv,int32_t maxv)
+{
+	if (v<minv) return minv;
+	if (v>maxv) return maxv;
+	return  v;
+}
+
+void draw_with_thickness(bitmap_rgb32 &bitmap,generic_plotter_device::Position pos,int thickness,uint32_t col)
+{
+	if (thickness<=0) return;
+	int left = clip(pos.x-thickness/2, 0, bitmap.width()-1);
+	int right = clip(pos.x+thickness/2, 0, bitmap.width()-1);
+	int top = clip(pos.y-thickness/2, 0, bitmap.height()-1);
+	int bottom = clip(pos.y+thickness/2, 0, bitmap.height()-1);
+	bitmap.plot_box(left, top, right-left+1, bottom-top+1, col);
+}
+
+
+void copybitmapregion(bitmap_rgb32 &dest,const bitmap_rgb32 &src,const generic_plotter_device::Position& destp, const generic_plotter_device::Position& srcp, int32_t width, int32_t height,const rectangle &cliprect )
+{
+	int32_t minx = std::max(std::max(0, destp.x-srcp.x),
+			      std::max(destp.x, cliprect.min_x)
+		); //assume arguments may be negative
+	int32_t miny = std::max(std::max(0, destp.y-srcp.y),
+			      std::max(destp.y, cliprect.min_y)
+		);
+	int32_t maxx = std::min(std::min(dest.width(), src.width()+destp.x-srcp.x),
+			      std::min(destp.x+width, cliprect.max_x+1)
+		);
+	int32_t maxy = std::min(std::min(dest.height(), src.height()+destp.y-srcp.y),
+			      std::min(destp.y+height, cliprect.max_y+1)
+		);
+	for (int32_t x = minx; x<maxx;x++)
+	{
+		for (int32_t y = miny; y<maxy;y++)
+		{
+			dest.pix(y,x) = src.pix(y-destp.y+srcp.y, x-destp.x+srcp.x);
+		}
+	}
```

Comment:
Seriously, you can do this by copying a bitmap with the clipping rectangle and offsets set correctly.

Reply:
done. 
copybitmap(dest, src, 0, 0, destp.x-srcp.x, destp.y-srcp.y, rectangle(minx, maxx, miny, maxy));